src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,175 @@
+/**
+* @file saga_impl.hpp
+* @author Prabhat Sharma
+*
+* Implementation of SAGA: A Fast Incremental Gradient Method With
+* Support for Non-Strongly Convex Composite Objectives.
+*
+* mlpack is free software; you may redistribute it and/or modify it under the
+* terms of the 3-clause BSD license.  You should have received a copy of the
+* 3-clause BSD license along with mlpack.  If not, see
+* http://www.opensource.org/licenses/BSD-3-Clause for more information.
+*/
+#ifndef MLPACK_CORE_OPTIMIZERS_SAGA_SAGA_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_SAGA_SAGA_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "saga.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename UpdatePolicyType, typename DecayPolicyType>
+SAGAType<UpdatePolicyType, DecayPolicyType>::SAGAType(
+    const double stepSize,
+    const size_t batchSize,
+    const size_t maxIterations,
+    const double tolerance,
+    const bool shuffle,
+    const UpdatePolicyType& updatePolicy,
+    const DecayPolicyType& decayPolicy,
+    const bool resetPolicy) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle),
+    updatePolicy(updatePolicy),
+    decayPolicy(decayPolicy),
+    resetPolicy(resetPolicy)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename UpdatePolicyType, typename DecayPolicyType>
+template<typename DecomposableFunctionType>
+double SAGAType<UpdatePolicyType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
```

Comment:
Can you use the `Function<>` class to provide any methods that `DecomposableFunctionType` might not have?  Take a look at `sgd_impl.hpp`, specifically the lines at the top:

```
  typedef Function<DecomposableFunctionType> FullFunctionType;
  FullFunctionType& f(static_cast<FullFunctionType&>(function));
```

Then, you could call `EvaluateWithGradient()` to get both the objective and the gradient at the same time.  (This can make a big difference!)

Reply:
I don' think it'll be useful here as I've had to calculate `Gradient()` and `Evaluate()` separately because `gradient` of only a random function is to be computed during one epoch whereas objective is to be computed for every `function`